go-api/utils/utils.go

```diff
@@ -0,0 +1,209 @@
+	if pgxError != nil {
+		return nil, nil, nil, pgxError
+	}
+	// redis connect
+	rdb := redis.NewClient(&redis.Options{
+		Addr: config["REDIS_HOST"].(string) + ":" + config["REDIS_PORT"].(string),
+	}).Conn()
+	_, rdbErr := rdb.Ping(context.Background()).Result()
+	if rdbErr != nil {
+		return nil, nil, nil, rdbErr
+	}
+	app := fiber.New(fiber.Config{
+		AppName:           "go-api v0.0.1",
+		EnablePrintRoutes: true,
+	})
+	app.Use(logger.New())
+	app.Use(cors.New())
+	app.Use(func(c *fiber.Ctx) error {
+		c.Locals("pgxConn", *pgxConn)
+		c.Locals("rdb", *rdb)
+		return c.Next()
+	})
+	return pgxConn, rdb, app, nil
+}
+
+func EncryptText(textToEncrypt string) (string, error) {
+	config := LoadEnvVars()
+	password := config["ENCRYPT_PASSWORD"].(string)
+	// Generate a random 16-byte IV
+	iv := make([]byte, 16)
+	if _, err := rand.Read(iv); err != nil {
+		fmt.Println("random 16byte iv generation failed")
+		return "", err
+	}
+
+	// Derive a 32-byte key using scrypt
+	key, err := scrypt.Key([]byte(password), []byte("salt"), 16384, 8, 1, 32)
+	if err != nil {
+		fmt.Println("scrypt failed")
+		return "", err
+	}
+
+	// Create a cipher using AES-256-CTR with the key and IV
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		fmt.Println("cipher generation failed")
+		return "", err
+	}
+	stream := cipher.NewCTR(block, iv)
+
+	// Encrypt the text
+	ciphertext := make([]byte, len(textToEncrypt))
+	stream.XORKeyStream(ciphertext, []byte(textToEncrypt))
+
+	// Combine the IV and ciphertext into a single string
+	encryptedText := hex.EncodeToString(iv) + hex.EncodeToString(ciphertext)
+
+	return encryptedText, nil
+}
+
+func DecryptText(encryptedText string) (string, error) {
+	config := LoadEnvVars()
+	password := config["ENCRYPT_PASSWORD"].(string)
+
+	// Extract the IV and ciphertext from the string
+	iv, err := hex.DecodeString(encryptedText[:32])
+	if err != nil {
+		return "", err
+	}
+	ciphertext, err := hex.DecodeString(encryptedText[32:])
+	if err != nil {
+		return "", err
+	}
+
+	// Derive the key using scrypt
+	key, err := scrypt.Key([]byte(password), []byte("salt"), 16384, 8, 1, 32)
+	if err != nil {
+		return "", err
+	}
+
+	// Create a decipher using AES-256-CTR with the key and IV
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return "", err
+	}
+	stream := cipher.NewCTR(block, iv)
+
+	// Decrypt the ciphertext
+	decryptedText := make([]byte, len(ciphertext))
+	stream.XORKeyStream(decryptedText, ciphertext)
+
+	return string(decryptedText), nil
+}
+
+func LoadEnvVars() map[string]interface{} {
+	return map[string]interface{}{
+		"DATABASE_URL":     os.Getenv("DATABASE_URL"),
+		"REDIS_HOST":       os.Getenv("REDIS_HOST"),
+		"REDIS_PORT":       os.Getenv("REDIS_PORT"),
+		"APP_PORT":         os.Getenv("APP_PORT"),
+		"TEST_MODE":        os.Getenv("TEST_MODE"),
+		"ENCRYPT_PASSWORD": os.Getenv("ENCRYPT_PASSWORD"),
+	}
+}
+
+func IsTestMode() bool {
```

Comment:
This function is called from places where we have access to `c.Locals`. `LoadEnvVars` are called inside of `Setup`. 

We could do the following inside of `Setup` and simplify or remove `IsTestMode`.

```go
config := LoadEnvVars()
testing := strconv.ParseBool(config["TEST_MODE"].(string))

app.Use(func(c *fiber.Ctx) error {
  c.Locals("pgxConn", *pgxConn)
  c.Locals("rdb", *rdb)
  c.Locals("testing", testing)
  return c.Next()
})
```

Reply:
I tried to implement your recommendation, but I'm afraid it might add more complexity. Parts where it reads `IsTestMode` are all in `routes` functions which doesn't have access fiber.ctx. If trying to use `Locals` then it would need more modification.

But still, if you insist I'll try to implement this

Reply:
after taking some time to think more deeply about this, I think it could be applied from controllers.